web/src/user_group_pill.ts

```diff
@@ -69,24 +71,24 @@ export function append_user_group(group: UserGroup, pill_widget: UserGroupPillWi
     pill_widget.clear_text();
 }
 
-export function get_group_ids(pill_widget: UserGroupPillWidget): number[] {
+export function get_group_ids(pill_widget: CombinedPillContainer): number[] {
     const items = pill_widget.items();
-    let group_ids = items.map((item) => item.id);
-    group_ids = group_ids.filter(Boolean);
-
-    return group_ids;
+    return items.flatMap((item) => item.type === "user_group" ? item.id : []);
 }
 
 export function filter_taken_groups(
     items: UserGroup[],
-    pill_widget: UserGroupPillWidget,
+    pill_widget: CombinedPillContainer,
 ): UserGroup[] {
     const taken_group_ids = get_group_ids(pill_widget);
     items = items.filter((item) => !taken_group_ids.includes(item.id));
     return items;
 }
 
-export function typeahead_source(pill_widget: UserGroupPillWidget): UserGroup[] {
+export function typeahead_source(pill_widget: CombinedPillContainer): UserGroupPillData[] {
     const groups = user_groups.get_realm_user_groups();
-    return filter_taken_groups(groups, pill_widget);
+    return filter_taken_groups(groups, pill_widget).map((user_group) => ({
+        ...user_group,
+        type: "user_group",
+    }));
```

Comment:
I think the refactoring to add these `type` fields may need to be done in a prep commit to make reviewing this be more feasible; as it is, the changes to `pill_typeahead.ts` are enough that GitHub doesn't even recognize it as a file rename.

Reply:
I can do that! Just wanted to check that this seemed like a reasonable strategy before taking the time to refactor, but I'll do that next.